frontend/lib/src/components/elements/DeckGlJsonChart/utils/colors.ts

```diff
@@ -0,0 +1,228 @@
+  SerializedColorValue?,
+  SerializedColorValue?,
+  SerializedColorValue?
+]
+
+type ObjectCallbackShape<T = unknown> = {
+  object: T
+  objectInfo: { index: number }
+}
+
+export type FillFunction<T = unknown> = (
+  object: ObjectCallbackShape<T>["object"],
+  objectInfo: ObjectCallbackShape<T>["objectInfo"]
+) => SerializedColorArray | SerializedColorValue
+
+type FillFunctionArgs<T = unknown> = ObjectCallbackShape<T> & {
+  originalFillFunction: FillFunction<T>
+}
+
+/**
+ * Retrieves the original color in a standardized format by evaluating functions
+ * if necessary.
+ */
+const getOriginalColor = ({
+  object,
+  objectInfo,
+  originalFillFunction,
+}: FillFunctionArgs): SerializedColorArray | null => {
+  const originalColor =
+    typeof originalFillFunction === "function"
+      ? originalFillFunction(object, objectInfo)
+      : originalFillFunction
+
+  if (Array.isArray(originalColor)) {
+    return [
+      originalColor[0] || 0,
+      originalColor[1] || 0,
+      originalColor[2] || 0,
+      originalColor[3] || 255,
+    ]
+  }
+
+  if (typeof originalColor === "string" && originalColor.startsWith("@@=")) {
+    // @see https://deck.gl/docs/api-reference/json/conversion-reference#functions-and-using-the--prefix
+
+    const evaluated = jsonConverter
+      .convert({ originalColor })
+      .originalColor(object)
+
+    return [
+      evaluated[0] || 0,
+      evaluated[1] || 0,
+      evaluated[2] || 0,
+      evaluated[3] || 255,
+    ]
+  }
+
+  return null
+}
+
+/**
+ * Calculates the original color with the applied opacity based on the selection
+ * mode.
+ *
+ * @param {boolean} isSelected - Indicates if the object is selected.
+ * @param {number} opacity - The opacity to apply.
+ * @param {object} object - The object for which the color is being calculated.
+ * @param {object} objectInfo - Additional information about the object.
+ * @param {Function} originalFillFunction - The function to get the original
+ * fill color.
+ * @returns {SerializedColorArray | null} The color with the applied opacity or
+ * null if the original color is not available.
+ */
+const getOriginalColorWithAppliedOpacity = ({
+  isSelected,
+  object,
+  objectInfo,
+  opacity,
+  originalFillFunction,
+}: {
+  isSelected: boolean
+  opacity: number
+} & FillFunctionArgs): SerializedColorArray | null => {
+  const originalColor = getOriginalColor({
+    object,
+    objectInfo,
+    originalFillFunction,
+  })
+
+  if (!originalColor) {
+    return null
+  }
+
+  let calculatedOpacity = 0
+
+  if (isSelected) {
+    // Some layers will have objects where the opacity is lower than the default
+    // selected opacity In this case, we want to use the higher opacity so that
+    // the differentiation between selected and unselected objects is more
+    // pronounced
+    calculatedOpacity = Math.max(
+      typeof originalColor[3] === "number"
+        ? originalColor[3]
+        : Number.MIN_SAFE_INTEGER,
+      opacity
+    )
```

Comment:
nit: I was a bit confused about the usage of `Number.MIN_SAFE_INTEGER` and `Number.MAX_SAFE_INTEGER` and needed a minute to understand why its used. Maybe its better to do it like:
 
```typescript
 Math.max(
      typeof originalColor[3] === "number"
        ? originalColor[3]
        : opacity,
      opacity
    )

// or...

calculatedOpacity = typeof originalColor[3] === "number" ?  Math.max(originalColor[3], opacity) : opacity

// or can't we just use 0 here since its the smallest possible opacity anyways? 
```